src/promptflow/promptflow/executor/_process_manager.py

```diff
@@ -0,0 +1,458 @@
+        self._executor_creation_func = executor_creation_func
+        self.context = multiprocessing.get_context("fork")
+
+    def new_process(self, i):
+        """
+        Create and start a new process using the 'fork' context.
+
+        :param i: Index of the input and output queue for the new process.
+        :type i: int
+        """
+        process = self.context.Process(
+            target=self._process_target_func,
+            args=(
+                self._executor_creation_func,
+                self._input_queues[i],
+                self._output_queues[i],
+                self._log_context_initialization_func,
+                self._current_operation_context,
+            ),
+            daemon=True,
+        )
+        process.start()
+        try:
+            self._process_info[i] = ProcessInfo(
+                index=i,
+                process_id=process.pid,
+                process_name=process.name,
+            )
+        except Exception as e:
+            bulk_logger.warning(
+                f"Unexpected error occurred while creating ProcessInfo for index {i} and process id {process.pid}. "
+                f"Exception: {e}"
+            )
+        return process
+
+    def end_process(self, i):
+        """
+        Terminates a specified process.
+
+        :param i: Index of the process to terminate.
+        :type i: int
+        """
+        try:
+            pid = self._process_info[i].process_id
+            process = psutil.Process(pid)
+            process.terminate()
+            process.wait()
+            self._process_info.pop(i)
+        except psutil.NoSuchProcess:
+            bulk_logger.warning(f"Process {pid} had been terminated")
+        except Exception as e:
+            bulk_logger.warning(
+                f"Unexpected error occurred while end process for index {i} and process id {process.pid}. "
+                f"Exception: {e}"
+            )
+
+    def restart_process(self, i):
+        """
+        Restarts a specified process by first terminating it then creating a new one.
+
+        :param i: Index of the process to restart.
+        :type i: int
+        """
+        self.end_process(i)
+        self.new_process(i)
+
+    def handle_signals(self, control_signal, i):
+        """
+        Handles control signals for processes, performing actions such as starting, ending,
+        or restarting them based on the received signal.
+
+        :param control_signal: The control signal indicating the desired action. It can be 'start', 'end', or 'restart'.
+        :type control_signal: str
+
+        :param i: Index of the process to control.
+        :type i: int
+        """
+        if control_signal == ProcessControlSignal.END:
+            self.end_process(i)
+        elif control_signal == ProcessControlSignal.RESTART:
+            self.restart_process(i)
+        elif control_signal == ProcessControlSignal.START:
+            self.new_process(i)
+
+
+def fork_processes_manager(
```

Comment:
The name seems confusing, should be something like "create_spawned_fork_process_manager?"

Reply:
fixed